src/ra_server.erl

```diff
@@ -518,51 +521,76 @@ handle_leader({PeerId, #append_entries_reply{success = false,
-                          {_, #{from := From}, _, after_log_append} ->
-                              [{reply, From,
-                                {wrap_reply, {Idx, Term}}} | Effects0];
-                          _ ->
-                              Effects0
-                      end,
+            Effects = after_log_append_reply(Cmd, Idx, Term, Effects0),
             {leader, State, Effects}
     end;
-handle_leader({commands, Cmds}, #{cfg := Cfg} =  State00) ->
+handle_leader({commands, Cmds}, #{cfg := #cfg{id = Self,
+                                              log_id = LogId} = Cfg,
+                                  cluster := Cluster} =  State00) ->
     %% TODO: refactor to use wal batch API?
     Num = length(Cmds),
-    {State0, Effects0} =
-        lists:foldl(fun(C, {S0, E0}) ->
-                            {ok, I, T, S, E} = append_log_leader(C, S0, E0),
-                            case C of
-                                {_, #{from := From}, _, after_log_append} ->
-                                    {S, [{reply, From,
-                                          {wrap_reply, {I, T}}} | E]};
-                                _ ->
-                                    {S, E}
+    case catch  lists:foldl(fun(C, {S0, E0}) ->
+                            case append_log_leader(C, S0, E0) of
+                                {ok, I, T, S, E} ->
+                                    {S, after_log_append_reply(C, I, T, E)};
+                                {not_appended, wal_down, _S, _E} = Result ->
+                                    throw(Result)
                             end
-                    end, {State00, []}, Cmds),
-    ok = incr_counter(Cfg, ?C_RA_SRV_COMMAND_FLUSHES, 1),
-    ok = incr_counter(Cfg, ?C_RA_SRV_COMMANDS, Num),
-    {State, _, Effects} = make_pipelined_rpc_effects(State0, Effects0),
+                    end, {State00, []}, Cmds) of
+        {State0, Effects0} ->
+            ok = incr_counter(Cfg, ?C_RA_SRV_COMMAND_FLUSHES, 1),
+            ok = incr_counter(Cfg, ?C_RA_SRV_COMMANDS, Num),
+            {State, _, Effects} = make_pipelined_rpc_effects(State0, Effects0),
+            {leader, State, Effects};
+        {not_appended, wal_down, State0, Effects} ->
+            ?WARN("~ts ~b commands NOT appended to log. Reason: wal_down",
+                  [LogId, length(Cmds)]),
+            CondEffs = case maps:to_list(maps:remove(Self, Cluster)) of
+                           [] -> [];
+                           [{PeerId, _} | _] ->
+                               [{next_event, cast, {transfer_leadership, PeerId}}]
+                       end,
 
-    {leader, State, Effects};
+            State = State0#{condition =>
+                            #{predicate_fun => fun wal_down_condition/2,
+                              transition_to => leader,
+                              %% TODO: make duration configurable?
+                              timeout => #{duration => 5000,
```

Comment:
We have seen enough false positives with 5s timeouts that I'd suggest making it 10s even though any hardcoded default will be wrong/suboptimal for certain installations.

Reply:
Actually, this should be enough time for `wal_down_condition/2` to observe a WAL process change, and even if it is observed at a later point, the resending can still do its job.

Reply:
yeah I think 5s is fine, it does affect - if we see ill effects we can make it configurable but I doubt it, worse case it  goes back to leader then have to go back to await_condition again.